### ISSUE TITLE
Support vsync

### ISSUE THREAD

**Issue by mbernat** [tonarino/just-gl#17]
It would be nice to get rid of the flicker in the colorful_triangle example.

I'm not sure there is a simple blocking call that waits for the next vsync. There is a DRM `page_flip` call that can both display a frame and also optionally produce a `PageFlip` event. We are already using this call but not only for presentation, not vsync.

One is presumably supposed to wait for those events somehow but I haven't found a way with the `drm` crate.
Existing implementations seem to be doing either a select on the DRM file descriptor [1], [2], or in case of slint they somehow use their event loop together with DRM call `receive_events` [3].

Alternatively, there is a DRM `wait_vblank` call that seems to have a blocking version but it also requires some "target sequence" parameter that has something to do with absolute or relative frame scheduling. I am not sure how those are supposed to work.

[1] https://gitlab.freedesktop.org/mesa/kmscube/-/blob/master/drm-legacy.c#L127
[2] https://github.com/dvdhrm/docs/blob/master/drm-howto/modeset-vsync.c#L591
[3] https://github.com/slint-ui/slint/blob/master/internal/backends/linuxkms/display/egldisplay.rs#L136

**Comment by strohel** [comment]
We did a quick pairing session with @mbernat today, but I didn't end up helping much.

There's at least some maybe-related docs.
- About vsync in OpenGL/EGL, though not sure if these are relevant as @mbernat says the problem may be at lower level (in libdrm):
  - https://www.khronos.org/opengl/wiki/Swap_Interval
  - https://registry.khronos.org/EGL/sdk/docs/man/html/eglSwapInterval.xhtml
  - https://registry.khronos.org/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml
- Kernel commit that introduced DRM_MODE_PAGE_FLIP_TARGET_ABSOLUTE/RELATIVE https://lists.freedesktop.org/archives/dri-devel/2016-August/114794.html
  Some of the comments added say:
  > NOTE: DRM_IOCTL_WAIT_VBLANK must still be used make sure the vblank sequence before the target one has passed before calling this ioctl.
- DRM_IOCTL_WAIT_VBLANK is called by `drmWaitVBlank()` from libdrm https://gitlab.freedesktop.org/mesa/drm/-/blob/main/xf86drm.c#L2675
- There whole "vblank test" in librdm: https://gitlab.freedesktop.org/mesa/drm/-/blob/main/tests/vbltest/vbltest.c
- https://manpages.debian.org/unstable/libdrm-dev/drm-kms.7.en.html#Page-Flipping says:
  > A call to drmModeSetCrtc(3) is executed immediately and forces the CRTC to use the new scanout buffer. If you want smooth-transitions without tearing, you probably use double-buffering. You need to create one framebuffer object for each buffer you use. You can then call drmModeSetCrtc(3) on the next buffer to flip. If you want to synchronize your flips with vertical-blanks, you can use drmModePageFlip(3) which schedules your page-flip for the next vblank.

@bschwind (or later) @PabloMansanet this could be a material for Friday.

**Comment by mbernat** [comment]
Thanks Matěj, these seem informative, I'll see if I find something helpful in there.

Meanwhile, we tried to understand page flips better with @PabloMansanet. We established that `receive_events` is blocking, which solves the waiting part of the problem. Unfortunately, our `page_flip` doesn't wait for vsyncs. It correctly informs us that a page flip happened via an event but that event arrives basically immediately.

The C & slint implementations are using all of this stuff basically the same way we are, in particular they don't do anything with the target sequences, so hopefully those aren't needed (but might still be worth a try).

**Comment by mbernat** [comment]
A nice resource in the form of comments under a blogpost about how DRM rendering is supposed to work, including info about CRTCs, scan-out, page flips, triple buffering and much more. By the same guy as one of the C implementations.

https://dvdhrm.wordpress.com/2012/09/13/linux-drm-mode-setting-api/

I'm still super confused how any of this relates to OpenGL, since there are various GL extensions that also do vsyncing. Likewise, in Vulkan you have swapchain extensions that can handle presentation. I suppose you can choose whether to do it via DRM or an extension?

**Comment by strohel** [comment]
> I'm still super confused how any of this relates to OpenGL, since there are various GL extensions that also do vsyncing. Likewise, in Vulkan you have swapchain extensions that can handle presentation. I suppose you can choose whether to do it via DRM or an extension?

Yeah I'm also cofused. But my guess would be that the "display system" (i.e. us in `just-gl`, Xorg, Wayland) would be responsible for translating the higher-level "OpenGL preferences" to lower-level (librdm) ones..? Feels like OpenGL implementation cannot know how to do a vblank on $arbitrary_display_platform.?

**Comment by mbernat** [comment]
> Yeah I'm also cofused. But my guess would be that the "display system" (i.e. us in `just-gl`, Xorg, Wayland) would be responsible for translating the higher-level "OpenGL preferences" to lower-level (librdm) ones..? Feels like OpenGL implementation cannot know how to do a vblank on $arbitrary_display_platform.?

Well, not an arbitrary platform but I suppose that's what the extensions are for: if you decide to implement it for a given hardware and driver, you can support anything you want in there.

But anyway, we have a much bigger problem: I just double checked and it turns out we're always using the same buffer for presentation, so screen tearing is unavoidable.

I assumed that either the GBM surface or EGL would take care of all this (glutin is supposed to be double buffered by default) but apparently not. On top of that, GBM surface API is super simple, so I don't think I can do anything with: https://docs.rs/gbm/latest/gbm/struct.Surface.html

There also doesn't seem to be a way to configure the number of buffers at surface creation (I just checked all the flags / usages / modifiers of various creation APIs).

I'm all out of ideas but I'll check what other implementations are doing with the surface.
